composer/callbacks/speed_monitor.py

```diff
@@ -41,84 +161,139 @@ class SpeedMonitor(Callback):
-
         # Add the new element
-        self.batch_wct_buffer.append(batch_wct)
-        self.batch_num_samples_buffer.append(batch_num_samples)
+        self.history_samples.append(state.timestamp.sample.value)
+        self.history_wct.append(state.timestamp.total_wct.total_seconds())
 
         # Log the throughput
-        if len(self.batch_num_samples_buffer) == self.window_size:
-            throughput = sum(self.batch_num_samples_buffer) / sum(self.batch_wct_buffer)
-            logger.log_metrics({'throughput/samples_per_sec': throughput})
+        if len(self.history_wct) == self.history_wct.maxlen:
+            world_size = dist.get_world_size()
+            elapsed_batches = len(self.history_samples) - 1
+            elapsed_samples = int(self.history_samples[-1]) - int(self.history_samples[0])
+            elapsed_wct = self.history_wct[-1] - self.history_wct[0]
+            batches_per_sec = elapsed_batches / elapsed_wct
+            samples_per_sec = elapsed_samples / elapsed_wct
+            dev_batches_per_sec = batches_per_sec / world_size
+            dev_samples_per_sec = samples_per_sec / world_size
+            logger.log_metrics({'throughput/batches_per_sec': batches_per_sec})
+            logger.log_metrics({'throughput/samples_per_sec': samples_per_sec})
+            logger.log_metrics({'throughput/device/batches_per_sec': dev_batches_per_sec})
+            logger.log_metrics({'throughput/device/samples_per_sec': dev_samples_per_sec})
+
+            # Compute token stats if dataloader.dataset has max_seq_len. Assumes no padding.
+            try:
+                max_seq_len = state.dataloader.dataset.max_seq_len  # type: ignore
+                # Only applicable to seq data / models
+                logger.log_metrics({'throughput/tokens_per_sec': samples_per_sec * max_seq_len})
+                logger.log_metrics({'throughput/device/tokens_per_sec': dev_samples_per_sec * max_seq_len})
+            except AttributeError:
+                pass
+
+            composer_model = state.model
+            if not isinstance(composer_model, ComposerModel):
+                composer_model = composer_model.module  # Pass through DDP wrapping
+            if hasattr(composer_model, 'flops_per_batch'):
+                model_flops_per_batch = composer_model.flops_per_batch  # type: ignore
+                flops_per_batch = None
+                if isinstance(model_flops_per_batch, (int, float)):
+                    flops_per_batch = model_flops_per_batch
+                elif isinstance(model_flops_per_batch, Callable):
+                    flops_per_batch = model_flops_per_batch(state.batch)
+                else:
+                    raise TypeError(f'flops_per_batch must be int, float, or callable accepting a batch and '
+                                    'returning an int or float. Instead, got {type(flops_per_batch)}.')
+                flops_per_sec = flops_per_batch * samples_per_sec
```

Comment:
I may be getting confused between samples and batches, but i think this should be `* batches_per_sec`. This also gives me pause on a design question. `flops_ber_batch` is hard thing to put a number on, because it depends on batch size, which the model does not know about.